solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.zookeeper.data.Stat;
+
+import static org.apache.solr.common.params.CommonParams.OMIT_HEADER;
+import static org.apache.solr.common.params.CommonParams.WT;
+import static org.apache.solr.response.RawResponseWriter.CONTENT;
+import static org.apache.solr.security.PermissionNameProvider.Name.COLL_READ_PERM;
+
+/**Exposes the content of the Zookeeper
+ * This is an expert feature that exposes the data inside the back end zookeeper.This API may change or
+ * be removed in future versions.
+ * This is not a public API. The data that is returned is not guaranteed to remain same
+ * across releases, as the data stored in Zookeeper may change from time to time.
+ */
+@EndPoint(path = "/cluster/zk/*",
+    method = SolrRequest.METHOD.GET,
+    permission = COLL_READ_PERM)
+public class ZkRead {
+  private final CoreContainer coreContainer;
+
+  public ZkRead(CoreContainer coreContainer) {
+    this.coreContainer = coreContainer;
+  }
+
+  @Command
+  public void get(SolrQueryRequest req, SolrQueryResponse rsp) {
+    String path = req.getPathTemplateValues().get("*");
+    if (path == null || path.isEmpty()) path = "/";
+    byte[] d = null;
+    try {
+      List<String> l = coreContainer.getZkController().getZkClient().getChildren(path, null, false);
+      if (l != null && !l.isEmpty()) {
+        String prefix = path.endsWith("/") ? path : path + "/";
+
+        rsp.add(path, (MapWriter) ew -> {
+          for (String s : l) {
+            try {
+              Stat stat = coreContainer.getZkController().getZkClient().exists(prefix + s, null, false);
+              ew.put(s, (MapWriter) ew1 -> {
+                ew1.put("version", stat.getVersion());
+                ew1.put("aversion", stat.getAversion());
+                ew1.put("children", stat.getNumChildren());
+                ew1.put("ctime", stat.getCtime());
+                ew1.put("cversion", stat.getCversion());
+                ew1.put("czxid", stat.getCzxid());
+                ew1.put("ephemeralOwner", stat.getEphemeralOwner());
+                ew1.put("mtime", stat.getMtime());
+                ew1.put("mzxid", stat.getMzxid());
+                ew1.put("pzxid", stat.getPzxid());
+                ew1.put("dataLength", stat.getDataLength());
+              });
+            } catch (Exception e) {
+              ew.put("s", Collections.singletonMap("error", e.getMessage()));
+            }
+          }
+        });
+
+      } else {
+        d = coreContainer.getZkController().getZkClient().getData(path, null, null, false);
+        if (d == null || d.length == 0) {
+          rsp.add(path, null);
+          return;
+        }
+
+        Map<String, String> map = new HashMap<>(1);
+        map.put(WT, "raw");
+        map.put(OMIT_HEADER, "true");
+        req.setParams(SolrParams.wrapDefaults(new MapSolrParams(map), req.getParams()));
```

Comment:
Why is this needed?

Reply:
This is important if we are ever writing out raw content that comes from ZK. We should not use any of the `ResponseWriter` in Solr

Reply:
Is this only the case for the situation where we write the data? what about the case of the child nodes? can those requests be written in other response formats? If this is the only format that's allowed, should we fail if another format is requested?

Reply:
no if you are requesting data , you should expect raw data. it will not honour the `wt` param

Reply:
yes, for list operations , you can get data in any format